libs/hmpb/src/marking.ts

```diff
@@ -0,0 +1,199 @@
+
+// NOTE: All values used in this module are in PDF user space `pt` units.
+
+const PT = 1;
+const IN = 72 * PT;
+const PX = IN / 96;
+
+const pageMargins = [
+  pageMarginsInches.left * IN,
+  pageMarginsInches.top * IN,
+] as const;
+
+const timingMarkCount = [34, 41] as const;
+const timingMarkSize = [
+  TIMING_MARK_DIMENSIONS.width * IN,
+  TIMING_MARK_DIMENSIONS.height * IN,
+] as const;
+
+const bubbleSize = [BUBBLE_WIDTH_PX * PX, BUBBLE_HEIGHT_PX * PX] as const;
+// We grow the mark slightly to help make up for slight variations in printing:
+const markSize = [bubbleSize[0] * 1.2, bubbleSize[1] * 1.2] as const;
+const markBorderRadius = markSize[1] * 0.5;
+const markSizeHalf = [markSize[0] * 0.5, markSize[1] * 0.5] as const;
+
+const writeInFontName = 'Roboto-Bold';
+const writeInFontSizeDefault = 12;
+const writeInFontSizeReduced = 10;
+
+/**
+ * Center of the top-left timing mark.
+ *
+ * [TODO] We may need to move this into the election definition, to support
+ * marking 3rd part ballots
+ */
+const gridOrigin = [
+  pageMargins[0] + 0.5 * timingMarkSize[0],
+  pageMargins[1] + 0.5 * timingMarkSize[1],
+] as const;
+
+/**
+ * Generates a PDF with bubble marks in the expected positions for the given
+ * ballot style and corresponding votes.
+ *
+ * Intended for printing over pre-printed HMPBs.
+ */
+export function generateMarkOverlay(
+  election: Election,
+  ballotStyleId: string,
+  votes: VotesDict
+): NodeJS.ReadableStream {
+  assert(
+    election.gridLayouts,
+    'cannot generate mark overlay for election with no grid layouts'
+  );
+
+  const layout = election.gridLayouts.find(
+    (l) => l.ballotStyleId === ballotStyleId
+  );
+  assert(layout, `no grid layout found for ballot style ${ballotStyleId}`);
+
+  const pageSizeIn = ballotPaperDimensions(election.ballotLayout.paperSize);
+  const pageSize = [pageSizeIn.width * IN, pageSizeIn.height * IN] as const;
+
+  const gridSize = [
+    pageSize[0] - 2 * pageMargins[0] - timingMarkSize[0],
+    pageSize[1] - 2 * pageMargins[1] - timingMarkSize[1],
+  ];
+
+  const doc = new PdfDocument({
+    bufferPages: true,
+    size: pageSize as [number, number],
+    compress: false,
+  });
+  doc.registerFont(writeInFontName, `${__dirname}/fonts/Roboto-Bold.ttf`);
+
+  let pageCount = 1; // First page is added automatically.
+
+  for (const pos of layout.gridPositions) {
+    let pageNumber = pos.sheetNumber * 2;
+    if (pos.side === 'front') pageNumber -= 1;
+
+    while (pageCount < pageNumber) {
+      doc.addPage();
+      pageCount += 1;
+    }
+
+    const contestVotes = votes[pos.contestId];
+    if (!contestVotes) continue;
+
+    const mark = markInfo(contestVotes, pos);
+    if (!mark) continue;
+
+    doc.switchToPage(pageNumber - 1); // Pages are 0-indexed in `pdfkit`.
+
+    const bubbleCenter = [
+      gridOrigin[0] + gridSize[0] * (pos.column / (timingMarkCount[0] - 1)),
+      gridOrigin[1] + gridSize[1] * (pos.row / (timingMarkCount[1] - 1)),
+    ];
```

Comment:
Is bubble center established by convention? Is there an easy way to share bubble render position logic between this renderer and the renderer for the base layer unmarked HMPB?

Reply:
Oh yeah, fair point - I was following the logic in the [renderer](https://github.com/votingworks/vxsuite/blob/f08c6b60e5abe542642138a980c013964a600707/libs/hmpb/src/render_ballot.tsx#L320). The different coordinate spaces might complicate a bit, but I could see value in unifying more of the logic/calculations across print/mark/interpret where possible. Will see what's feasible in future iterations

Reply:
Hmm yeah seems like we could benefit from some shared logic especially since we've been through a few ballot layout reworks already. But agree it's probably too early to nail that down now.